### PR TITLE
Exploded scripts

### PR DESCRIPTION
**Pull request by ahgittin**
Starts expanding the scripts so we have piece-by-piece instructions.
See `all.sh` and the `TODO` items in there.

Builds on #1

### CODE REVIEW

**Comment by ahgittin** [comment]
@rdowner i've cleaned this up, merged changes, and expanded it further, with 4 clear steps and instructions in `all.sh`.  i'm confirming now that it's working but wanted to get early feedback.

answers to your comments below.

---

> > why not use a whitelist for brooklyn-server ?
> Basically as an insurance policy - brooklyn-server is everything that the others are not, so we can be sure that every historical folder will be in at least one of the new repos. If a historical folder name is missed out of one of the repo's whitelists, we know that it will still be in brooklyn-server.

good idea.  i've implemented this a different way, when we make the full whitelist i generate an additional list of unclaimed files and append that to `server-whitelist.full.txt`.  this way we can inspect the full whitelists and unclaimed files.

> > should we drop the M and rc branches? that would speed things up considerably, and i don't think we need them. we will still have the incubator repo.
> Actually it's not much of a speed penalty - going from master-only to master-with-all-release branches changed the number of commits from around 12,600 to 13,000 - a pretty marginal increase.
> > and drop 0.4 and 0.5 in favor of 0.x.0...?
> Do you mean rename the branches instead of drop? If so that's an easy thing to do.

i had wanted a simpler set of initial branches.  we have both 0.4 and 0.4.0, with the latter being the former plus the version update.  i've set it in `env.sh` as

    branches="0.4.0 0.5.0 0.6.0 0.7.0-incubating 0.8.0-incubating"

i'm not entirely sure it makes sense to keep any branches, as they won't compile and so will confuse people -- however it might be useful to see when branches were cut.  on balance i'm probably +0 to having no branches -- wdyt?

> > another comment, i think we should restructure in incubator-brooklyn
> When I tried this, when you look at the history in the GitHub "commits" page, it stopped on the rename. So the new repository looked at first glance as if it had no history. It was still all there of course, but only if you drilled down into it.

can you elaborate?  this would be a good reason not to do what i propose.  but at https://github.com/ahgittin/incubator-brooklyn/tree/reorg i see the history fine so i'm not clear what you mean.

> > also we need to add a step which removes big files
> The script has a prototype for doing this, commented out at the top (commented out because it involved another 15-minutes pass over the repo, and I was more interested in debugging the split section).

missed this; sorry.  i've applied it in `2-...sh`

> Could we use `$( cat library-whitelist.txt )` for this and all the other cases? Less likely for mistakes if that information is not duplicated.

+1, done

> `git mv` to ensure that git correctly records the metadata? Git sometimes gets it wrong when left to figure out moves on its own.

+1, done

File: 2-clean-history.sh [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+    examples/simple-messaging-pubsub/src/main/resources/je-5.0.34.jar
+    examples/simple-nosql-cluster/src/main/resources/cumulusrdf-0.6.1-pre.jar
+    examples/simple-nosql-cluster/src/main/resources/cumulusrdf.war
+    examples/simple-nosql-cluster/src/main/resources/cumulusrdf.war
+    sandbox/examples/src/main/resources/gemfire/springtravel-datamodel.jar
+    sandbox/examples/src/main/resources/swf-booking-mvc.war' HEAD
+
+popd
+
+echo size after cleaning history:
```
Comment (rdowner) [inline 2-clean-history.sh]: A few extra steps are needed to properly clean the history - see this extract from the git filter-branch help page:

> - Remove the original refs backed up by git-filter-branch: say `git for-each-ref --format="%(refname)" refs/original/ | xargs -n 1 git update-ref -d`.
> - Expire all reflogs with `git reflog expire --expire=now --all`.
> - Garbage collect all unreferenced objects with `git gc --prune=now` (or if your git-gc is not new enough to support arguments to --prune, use `git repack -ad; git prune` instead).

**Comment by rdowner** [comment]
Thanks @ahgittin - happy to merge this.

A few new comments and replies-to-comments:

* See the comment on `2-clean-history.sh` for a bit of git magic to properly clean the deleted objects (`git filter-branch` keeps backups of the original refs, so the objects stick around and the repository doesn't actually shrink in size on the local machine)(but when pushed to remote, we wouldn't push the backup refs, so the remote repo *would* be smaller). Not a deal-breaker as it still all works in the end, but worth doing if you want to see a better "this much space saved" stat :-)
* Happy to go with the smaller set of branches. Do we want to delete any tags from the result, too? (they are all ported by default)
* History view in GitHub: the history for the monolithic repository looks OK, but when you go to one of the modular repo directories, all you see in the history is [this](https://github.com/ahgittin/incubator-brooklyn/commits/reorg/brooklyn-dist). It may not be a big enough problem to worry about though, wdyt?

**Comment by rdowner** [comment]
ObDisclaimer: I've looked at the scripts but haven't actually run them :-O

**Comment by ahgittin** [comment]
> See the comment on 2-clean-history.sh for a bit of git magic to properly clean the deleted objects 

applied, new PR coming soon

> Happy to go with the smaller set of branches. Do we want to delete any tags from the result, too? (they are all ported by default)

tags, hmm, yes probably.  i'm not very familiar w git tags tho!

> History view in GitHub: the history for the monolithic repository looks OK, but when you go to one of the modular repo directories, all you see in the history is this. It may not be a big enough problem to worry about though, wdyt?

gotcha, that's expected, and not a problem in my view.  and assuming we do a `git mv brooklyn-server/{*,.??*} .` as the first commit in each of the repos, you won't actually be going in to the `brooklyn-server` dir.

**Pull request merged**